devito/types/dimension.py

```diff
@@ -181,6 +181,14 @@ def min_name(self):
     def max_name(self):
         return "%s_M" % self.name
 
+    @property
+    def indirect(self):
+        return False
+
+    @property
+    def index(self):
+        return self if self.indirect is True else getattr(self, 'parent', self)
```

Comment:
A class shouldn't have any knowledge of its subclasses

Here you're trying to access the parent, which is subclass-only stuff, otherwise you wouldn't use `getattr` (which should always be avoided unless in very specific cases)

the way to go is:

* in the superclass: systematically `return self `
* in DerivedDimension: use the if-else, then no need for getattr

Reply:
Except we have dimensions such as `CustomDimension` that are not `DerivedDimension` but can be indirect and have a parent in some cases.